### PR TITLE
8263002: Remove CDS MiscCode region

### PR DESCRIPTION
**Pull request by iklam**
The CDS MiscCode region is used for:
(a) C++ vtables
(b) Method trampolines

(a) can be moved to the ReadWrite region
(b) were introduced in JDK-8145221 so we can delay writing into Methods. This was intended to improve copy-on-write sharing to reduce memory footprint. However, this hasn't been shown to have any significant effect (footprint of metadata usually is much smaller than the Java heap), and introduces a lot of complexity in the HotSpot code.

Removing (b) will make it easier to implement JDK-8026297 (Generating AdapterHandlerEntry during CDS dump), which will further improve start-up time.

============
Other benefits of removing the MiscCode region:

- We no longer have a read/write/executable region. This address the concern in JDK-8262922.
- We can enable CDS on macOS/AArch64, which does not allow read/write/executable regions. (JDK-8253795)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263002](https://bugs.openjdk.java.net/browse/JDK-8263002): Remove CDS MiscCode region

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 1c6ac95e9f66a740a43a75ad428f011c3cf7f9c4
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to a3e4f25e2928b89522d73bbf9170713c9b73acc8

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2861/head:pull/2861`
`$ git checkout pull/2861`

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
wow.  Looks good to me.

File: src/hotspot/share/memory/archiveBuilder.cpp [hunk 1]
```diff
@@ -848,9 +827,13 @@ class RelocateBufferToRequested : public BitMapClosure {
       ArchivePtrMarker::ptrmap()->clear_bit(offset);
     } else {
       if (STATIC_DUMP) {
-        assert(_builder->is_in_buffer_space(*p), "old pointer must point inside buffer space");
-        *p += _buffer_to_requested_delta;
-        assert(_builder->is_in_requested_static_archive(*p), "new pointer must point inside requested archive");
+        if (!_builder->is_in_buffer_space(*p)) {
+          tty->print_cr("ohashii %p %p", p, *p);
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/memory/archiveBuilder.cpp]: This was debug code that I left behind by mistake. I've removed it.

File: src/hotspot/share/memory/cppVtables.cpp [hunk 2]
```diff
@@ -215,7 +215,7 @@ CppVtableInfo** CppVtables::_index = NULL;
 char* CppVtables::dumptime_init() {
   assert(DumpSharedSpaces, "must");
   size_t vtptrs_bytes = _num_cloned_vtable_kinds * sizeof(CppVtableInfo*);
-  _index = (CppVtableInfo**)ArchiveBuilder::current()->mc_region()->allocate(vtptrs_bytes);
+  _index = (CppVtableInfo**)ArchiveBuilder::current()->rw_region()->allocate(vtptrs_bytes);
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/memory/cppVtables.cpp]: The vtables contain function pointers and need to be updated at runtime to match the latest location of libjvm.so, so they need to be R/W.

**Review by dholmes-ora — APPROVED** [review]
Hi Ioi,

This looks really good from a code simplification point of view!

I can't claim to fully understand all the code involved though.

Thanks,
David

**Comment by kelthuzadx** [comment]
Hi @iklam, sorry to bother you, I found there are some comments that still refer to [lingering mc regions](https://github.com/openjdk/jdk/compare/master...kelthuzadx:residual_mc_region).

**Comment by iklam** [comment]
> Hi @iklam, sorry to bother you, I found there are some comments that still refer to [lingering mc regions](https://github.com/openjdk/jdk/compare/master...kelthuzadx:residual_mc_region).

Thanks @kelthuzadx , I have filed [JDK-8263998](https://bugs.openjdk.java.net/browse/JDK-8263998) (Remove mentions of mc region in comments)

**Pull request closed without merging**